File: typescript/src/worker.ts
#!/usr/bin/env node
/**
 * Standalone worker entry point for loop scheduler
 */

import * as path from 'path';
import { LoopManager } from './loop-manager.js';
import { SkillManager } from './skill-manager.js';
import { StateManager } from './state-manager.js';
import { LoopLogger } from './logger.js';
import { LoopExecutor } from './loop-executor.js';
import { LoopScheduler } from './scheduler.js';

async function main() {
  const workspaceRoot = process.cwd();
  const loopDir = path.join(workspaceRoot, '.loop');

  const loopManager = new LoopManager(loopDir);
  const skillManager = new SkillManager(path.join(loopDir, 'skills'));
  const stateManager = new StateManager(path.join(loopDir, 'state'));
  const logger = new LoopLogger(path.join(loopDir, 'logs'));
  const executor = new LoopExecutor(
    workspaceRoot,
    loopManager,
    skillManager,
    stateManager,
    logger
  );
  const scheduler = new LoopScheduler(loopManager, logger, (name) => executor.queueRun(name));

  await logger.info('worker_started', undefined, { workspace: workspaceRoot });
  scheduler.start();

  process.on('SIGINT', async () => {
    await scheduler.stop();
    await logger.info('worker_stopped');
    process.exit(0);
  });
}

main().catch((err) => {
  console.error('Worker error:', err);
  process.exit(1);
});

File: python/src/loop_engineering_mcp/worker.py
"""Standalone worker entry point for loop scheduler."""

import asyncio
import sys
from pathlib import Path

from .loop_manager import LoopManager
from .skill_manager import SkillManager
from .state_manager import StateManager
from .logger import LoopLogger
from .loop_executor import LoopExecutor
from .scheduler import LoopScheduler


async def run_worker():
    """Run the loop scheduler as a standalone background worker."""
    workspace_root = Path.cwd()
    loop_dir = workspace_root / ".loop"
    loop_dir.mkdir(exist_ok=True)

    loop_manager = LoopManager(loop_dir)
    skill_manager = SkillManager(loop_dir / "skills")
    state_manager = StateManager(loop_dir / "state")
    logger = LoopLogger(loop_dir / "logs")
    executor = LoopExecutor(
        workspace_root, loop_manager, skill_manager, state_manager, logger
    )
    scheduler = LoopScheduler(loop_manager, logger, run_callback=executor.queue_run)

    logger.info("worker_started", workspace=str(workspace_root))
    scheduler.start()

    try:
        while True:
            await asyncio.sleep(3600)
    except asyncio.CancelledError:
        pass
    finally:
        await scheduler.stop()
        logger.info("worker_stopped")


def main():
    """CLI entry point for loop-engineering-worker."""
    try:
        asyncio.run(run_worker())
    except KeyboardInterrupt:
        print("\nShutting down loop worker...", file=sys.stderr)
        sys.exit(0)


if __name__ == "__main__":
    main()
